lexer/lex.go

```diff
@@ -1,13 +1,415 @@
+
+	tokens := make(chan Tokval)
+	
+	go func() {
+	
+		decodedCode := code.Runes()
+		currentState := newLexer(decodedCode).initialState
+		
+		for currentState != nil {
+			token, newState := currentState()
+			tokens <- token
+			currentState = newState
+		}
+		
+		close(tokens)
+	}()
+
+	return tokens
+}
+
+type lexer struct {
+	code []rune
+	position uint
+	line uint
+	column uint
+}
+
+type lexerState func() (Tokval, lexerState)
+
+func newLexer(code []rune) *lexer {
+	return &lexer {code:code, line: 1, column: 1}
+}
+
+func (l *lexer) initialState() (Tokval, lexerState) {
+
+	if l.isEOF() {
+		return EOF, nil
+	}
+		
+	if l.isInvalidRune() {
+		return l.illegalToken()
+	}
+	
+	if l.isPlusSign() {
+		// TODO: handle ++
+		return l.token(token.Plus), l.initialState
+	}
+	
+	if l.isMinusSign() {
+		// TODO: handle --
+		return l.token(token.Minus), l.initialState
+	}
+		
+	if l.isNumber() {
+		l.fwd()
+		return l.numberState()
+	}
+		
+	if l.isDot() {
+		l.fwd()
+		allowExponent := true
+		allowDot := false
+		return l.decimalState(allowExponent, allowDot)
+	}
+	
+	if l.isRightParen() {
+		return l.token(token.RParen), l.initialState
+	}
+	
+	if l.isComma() {
+		return l.token(token.Comma), l.initialState
+	}
+	
+	if l.isDoubleQuote() {
+		l.fwd()
+		return l.stringState()
 	}
-)
\ No newline at end of file
+		
+	return l.identifierState()
+}
+
+func (l *lexer) stringState() (Tokval, lexerState) {
+	// TODO: handle newlines
+	
+	for !l.isEOF() && !l.isDoubleQuote() {
+		l.fwd()
+	}
+	
+	if l.isEOF() {
+		return l.illegalToken()
+	}
+	
+	return l.stringToken(), l.initialState
+}
+
+func (l *lexer) numberState() (Tokval, lexerState) {
+
+	if l.isEOF() {
+		return l.token(token.Decimal), l.initialState
+	}
+	
+	if l.isHexStart() {
+		l.fwd()
+		
+		if l.isEOF() {
+			return l.illegalToken()
+		}
+		
+		return l.hexadecimalState()
+	}
+	
+	allowExponent := true
+	allowDot := true
+	return l.decimalState(allowExponent, allowDot)
+}
+
+func (l *lexer) illegalToken() (Tokval, lexerState) {
+	return Tokval{
+		Type: token.Illegal,
+		Value: newStr(l.code),
+	}, nil
+}
+
+func (l *lexer) identifierState() (Tokval, lexerState) {
+
+	for !l.isEOF() {
+		if l.isDot() {
+			l.bwd()
+			return l.token(token.Ident), l.accessMemberState
+		}
+	
+		if l.isLeftParen() {
+			l.bwd()
+			return l.token(token.Ident), l.leftParenState
+		}
+		l.fwd()
+	}
+		
+	return l.token(token.Ident), l.initialState
+}
+
+func (l *lexer) leftParenState() (Tokval, lexerState) {
+	return l.token(token.LParen), l.initialState
+}
+
+func (l *lexer) startIdentifierState() (Tokval, lexerState) {
+
+	if l.isEOF() {
+		return EOF, nil
+	}
+	
+	if l.isNumber() {
+		return l.illegalToken()
+	}
+	if l.isDot() {
+		return l.illegalToken()
+	}
+	return l.identifierState()
+}
+
+func (l *lexer) accessMemberState() (Tokval, lexerState) {
+	return l.token(token.Dot), l.startIdentifierState
+}
+
+func (l *lexer) hexadecimalState() (Tokval, lexerState) {
+
```

Comment:
didn't liked the idea of: `l.accept(hexChars)` ?

Reply:
not necessarily, it naturally evolved in my mind like this but we could refactor later =D